decisions/2019/10/09-fe-decision-board.md

```diff
@@ -0,0 +1,226 @@
+
+### Unified form handling #25
+
+**Conclusion**: Going forward with
+[`dhis2/ui-forms`](https://github.com/dhis2/ui-core/pull/430) as is.
+Work on it is currently on hold as some required components are being
+re-worked and developed for the next major version of
+[`dhis2/ui-core`](https://github.com/dhis2/ui-core/pull/430), so now is
+a _great time_ to go through `ui-forms` and get involved!
+
+### Slow down releases a bit, improve automation
+
+**Conclusion**: The proposal was rejected. This was in part due to
+wishes to not split the manual workflow in two by having e.g.
+squash-merges to `develop` and merges to `master`.
+
+On hindsight there are multiple issues conflated in this topic:
+
+- Allow Dependabot **automatic merges** of certain dependencies to handle
+  the avalanche of dependency updates.
+
+- Allow a developer to **defer the automatic release system** for a
+  short amount of time to bundle commits into the same release.
+
+- Allow the developer to **use the same workflow everywhere** so she
+  doesn't have to keep track of what to do when in 40+ repos.
+
+Going through them one by one:
+
+#### Automatic merges
+
+**Trial**: Allow Dependabot to merge a sub-set of dependencies to
+`master` in a low-risk repository:
+[cli-style](https://github.com/dhis2/cli-style)
+
+The up-to-date automerged dependencies can be seen in
+[`cli-style/.dependabot/config.yml`](https://github.com/dhis2/cli-style/blob/master/.dependabot/config.yml)
+at any time. At the time of writing the strategy is to automatically
+merge any **minor** updates to any `@dhis2` scoped package and merge all
+**patch** versions for known security fixes. Others will be manually
+handled.
+
+Will run the trial for 1 week in **cli-style** before rolling out to
+**ui-widgets**. If everything keeps going all right we will identify
+another candidate from the [Dependabot enabled
+repos](https://app.dependabot.com/accounts/dhis2).
+
+#### Defer the automatic release system
+
+There are a few options of varying complexity:
+
+- The proposed workflow in the agenda.
+
+- Bundle multiple changes on the same branch before merging to `master`.
+
+- Using `[skip ci]` in a commit message to, well, skip CI, which will
+  also skip the `semantic-release` process.
+
+- Implementing another commit message tag, e.g. `[defer release]` to
+  still run the CI checks, but skip the `semantic-release` process. [A PR
+  for it has been posted](https://github.com/dhis2/cli/pull/161).
+
+The `[skip ci]` and `[defer release]` options are different variations
+of the same, so we will go ahead with one of them. If the defer-release
+PR is accepted, we will use that. If not, we will use skip-ci.
+
+#### Use the same workflow everywhere
+
+Currently we use "squash-merge" as our default strategy, and we also use
+`next` branches to merge multiple fixes/features to `master` at the same
+time.
+
+There is a significant downside of this in terms of how the changelogs
+are generated.
```

Comment:
If we want to keep the option to use next-branches, why can't we just fast-forward merge in cases where we use a next-branch, and keep squash-merging for features that are merged directly from feature-branches?

Reply:
The main reason against having two merge strategies is that it opens up error opportunities.

The standard merge in GitHub is never a fast-forward, so that is out.

We don't allow pushing to `master`, so we cannot do a fast-forward merge locally and then push.

That leaves squash-merge and rebase-merge, both of these are fast-forward merges except one squashes the commits and one doesn't.

A squash-merge from a feature-branch to a `next` branch is OK.

A squash-merge from a feature-branch to `master` is _might_ be OK.  It might also not be OK at all.

If there are two commits, the first is a breaking change and the second is a fix, the risk when squash-merging that to master is that the fix is used as the final commit message, and then we have a breaking change releases as a patch-version.

A problem with the squash-merge is that we cannot enforce the final commit message defined in GitHub to follow our commit message standards. If you want to put in "FOOBAR YO THIS COMMIT IS GR8" there's nothing stopping you from doing so.

If we rebase-merge from a feature-branch to `master` then all commits land on master and the changelog and version bump can be automatically detected. The merging developer doesn't have to worry too much about it, as she cannot define a commit message when rebase-merging, and we enforce all the commit messages to follow the standard before pushing.

If you keep your commits clean-ish, there's no problem doing a rebase-merge from a feature-branch to a next-branch either, so you do not _have_ to squash your commits if you don't need to.

The general consensus during the meeting was that we would prefer to have a single workflow and merge strategy so that we do not have to think "is this safe to squash-merge, or do I have to rebase-merge it?" and can just merge if all the status checks are OK.

The rebase-merge strategy has different drawbacks than squash-merge, that is true. Though when looking at how our entire process works, the rebase-merge has less cons, and some nice pros that we don't have with a squash-merge.

Let's try it out in a few places to see how it works out.

Reply:
>  If you want to put in "FOOBAR YO THIS COMMIT IS GR8" there's nothing stopping you from doing so.

I've actually had a typo in the semantic type before, so it's a nice safety net indeed.

Overall, I agree. I just wanted to raise an alternative and other cons to see if we had considered those.

Reply:
That's great, much appreciated! That's exactly what this is for. 👍